dvc/remote/gdrive/pydrive.py

```diff
@@ -0,0 +1,93 @@
+import os
+
+from dvc.remote.gdrive.utils import TrackFileReadProgress, FOLDER_MIME_TYPE
+
+
+class RequestBASE:
+    def __init__(self, drive):
+        self.drive = drive
+
+    def execute(self):
+        raise NotImplementedError
+
+
+class RequestListFile(RequestBASE):
+    def __init__(self, drive, query):
+        super(RequestListFile, self).__init__(drive)
+        self.query = query
+
+    def execute(self):
+        return self.drive.ListFile(
+            {"q": self.query, "maxResults": 1000}
+        ).GetList()
+
+
+class RequestListFilePaginated(RequestBASE):
+    def __init__(self, drive, query):
+        super(RequestListFilePaginated, self).__init__(drive)
+        self.query = query
+        self.iter = None
+
+    def execute(self):
+        if not self.iter:
+            self.iter = iter(
+                self.drive.ListFile({"q": self.query, "maxResults": 1000})
+            )
+        return next(self.iter, None)
+
+
+class RequestUploadFile(RequestBASE):
+    def __init__(
+        self, args, no_progress_bar=True, from_file="", progress_name=""
+    ):
+        super(RequestUploadFile, self).__init__(args["drive"])
+        self.title = args["title"]
+        self.parent_id = args["parent_id"]
+        self.mime_type = args["mime_type"]
+        self.no_progress_bar = no_progress_bar
+        self.from_file = from_file
+        self.proress_name = progress_name
+
+    def upload(self, item):
+        with open(self.from_file, "rb") as from_file:
+            if not self.no_progress_bar:
+                from_file = TrackFileReadProgress(self.proress_name, from_file)
+            if os.stat(self.from_file).st_size:
+                item.content = from_file
+            item.Upload()
+
+    def execute(self):
```

Comment:
is atomic? e.g. if upload fails after file create, will we potentially end up with a broken remote? should we first create a temporary file like we do in other places, and then move it in one operation?

Reply:
`Upload` should be atomic, I can't find direct mention of it in docs, but there is a list of general recommendations https://developers.google.com/drive/api/v2/manage-uploads#errors

Reply:
even though upload itself is atomic what would happen if we interrupt it during it. Will the item created with CreateFile before be still visible? So, my questions is not about upload per se, more about our _ RequestUploadFile ::execute_ implementation if it makes sense.

Reply:
> Will the item created with CreateFile before be still visible?

`CreateFile` creates only local object and doesn't create anything at remote drive.
`Upload` is the only func to communicate with remote API and shouldn't cause appearing of corrupted files in case of error.

Reply:
got it. Thanks @MaxRis 👍

Reply:
Remote item's `id` becomes available only after call of `Upload`.